In [1]:
import pandas as pd
import math

In [2]:
df_main = pd.read_json("./experiment_queries_execution_final_methods.json", orient="records")
df = df_main[df_main["method_id"].str.contains("boost")]
    

In [3]:
# Number of methods
df["method_id"].unique()


array(['single_elastic_bm25_boost_0.5_cross_encoder',
       'single_qdrant_sparse_boost_0.5_cross_encoder',
       'single_qdrant_dense_boost_0.5_cross_encoder',
       'tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder',
       'triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder'],
      dtype=object)

# Overall

In [4]:
df_thin = df[["method_id", "rr", "time"]]
df_thin.head()

method_id   rr      time
7540  single_elastic_bm25_boost_0.5_cross_encoder  0.0  1.079723
7541  single_elastic_bm25_boost_0.5_cross_encoder  0.0  1.087204
7542  single_elastic_bm25_boost_0.5_cross_encoder  0.0  1.295412
7543  single_elastic_bm25_boost_0.5_cross_encoder  0.0  1.191521
7544  single_elastic_bm25_boost_0.5_cross_encoder  0.0  0.971411

In [5]:
df_overall =  df_thin.groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False)
print(df_overall.to_string())

                                                                                       method_id        rr      time
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.620679  1.339842
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.620636  1.329294
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.620092  1.337697
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.613774  1.252900
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.486331  1.353622


### Overall only tuples

In [6]:
df_overall_tuple_only = df_overall[df_overall["method_id"].str.contains("tuple")]
df_overall_tuple_only

method_id        rr      time
4  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...  0.620679  1.339842

### Overall only triples

In [7]:
df_overall_triple_only = df_overall[df_overall["method_id"].str.contains("triple")]
print(df_overall_triple_only.to_string())

                                                                                       method_id        rr      time
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.620092  1.337697


# Overall per category



In [8]:
df_thin_per_category = df[["method_id", "class_category", "rr", "time"]]
df_thin_per_category

method_id  \
7540         single_elastic_bm25_boost_0.5_cross_encoder   
7541         single_elastic_bm25_boost_0.5_cross_encoder   
7542         single_elastic_bm25_boost_0.5_cross_encoder   
7543         single_elastic_bm25_boost_0.5_cross_encoder   
7544         single_elastic_bm25_boost_0.5_cross_encoder   
...                                                  ...   
15075  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15076  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15077  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15078  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15079  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   

                    class_category     rr      time  
7540            instances_100_1000  0.000  1.079723  
7541            instances_100_1000  0.000  1.087204  
7542            instances_100_1000  0.000  1.295412  
7543            instances_100_1000  0.000  1.191521  
7544            instances_100_1000  0.000  0.971411  
...                            ...    ...       ...  
15075  instances_100000_1000000000  1.000  1.166308  
15076  instances_100000_1000000000  1.000  1.448362  
15077  instances_100000_1000000000  0.200  1.218570  
15078  instances_100000_1000000000  0.125  1.189481  
15079  instances_100000_1000000000  0.500  1.167621  

[7540 rows x 4 columns]

In [9]:
df_overall_per_category = df_thin_per_category.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category","rr"], ascending=False)
print(df_overall_per_category.to_string())

                                                                                        method_id               class_category        rr      time
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder           instances_100_1000  0.606961  1.307137
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder           instances_100_1000  0.606630  1.315498
15                                                    single_qdrant_dense_boost_0.5_cross_encoder           instances_100_1000  0.605858  1.303630
23                                                   single_qdrant_sparse_boost_0.5_cross_encoder           instances_100_1000  0.605340  1.232038
7                                                     single_elastic_bm25_boost_0.5_cross_encoder           instances_100_1000  0.506691  1.340534
14                                                    single_qdrant_dense_boost_0.5_cross_encoder         instances_10

### Only instance categories

In [10]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("instances")].to_string())

                                                                                        method_id               class_category        rr      time
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder           instances_100_1000  0.606961  1.307137
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder           instances_100_1000  0.606630  1.315498
15                                                    single_qdrant_dense_boost_0.5_cross_encoder           instances_100_1000  0.605858  1.303630
23                                                   single_qdrant_sparse_boost_0.5_cross_encoder           instances_100_1000  0.605340  1.232038
7                                                     single_elastic_bm25_boost_0.5_cross_encoder           instances_100_1000  0.506691  1.340534
14                                                    single_qdrant_dense_boost_0.5_cross_encoder         instances_10

### Only ancestors cetegories

In [11]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("ancestors")].to_string())

                                                                                        method_id            class_category        rr      time
9                                                     single_qdrant_dense_boost_0.5_cross_encoder  ancestors_180_1000000000  0.786538  1.375448
25  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  ancestors_180_1000000000  0.776843  1.390154
33                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  ancestors_180_1000000000  0.776843  1.381017
17                                                   single_qdrant_sparse_boost_0.5_cross_encoder  ancestors_180_1000000000  0.674359  1.322899
1                                                     single_elastic_bm25_boost_0.5_cross_encoder  ancestors_180_1000000000  0.557692  1.486132
16                                                   single_qdrant_sparse_boost_0.5_cross_encoder         ancestors_150_180  0.663329  1

### Only children

In [12]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("children")].to_string())

                                                                                        method_id              class_category        rr      time
3                                                     single_elastic_bm25_boost_0.5_cross_encoder       children_10000_100000  0.582922  1.301950
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder       children_10000_100000  0.551052  1.263654
11                                                    single_qdrant_dense_boost_0.5_cross_encoder       children_10000_100000  0.550770  1.307097
35                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder       children_10000_100000  0.550603  1.293033
19                                                   single_qdrant_sparse_boost_0.5_cross_encoder       children_10000_100000  0.544494  1.193731
10                                                    single_qdrant_dense_boost_0.5_cross_encoder  children_100000_100000000

# Overall - only mechanic queries

In [13]:
df_only_mech_queries = df[df["query_id"].str.startswith("mech")]
df_only_mech_queries

method_id  user_id  class_id  \
8212         single_elastic_bm25_boost_0.5_cross_encoder        0  20825628   
8213         single_elastic_bm25_boost_0.5_cross_encoder        0  20825628   
8214         single_elastic_bm25_boost_0.5_cross_encoder        0  20825628   
8215         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
8216         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
...                                                  ...      ...       ...   
15075  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15076  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15077  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15078  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15079  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   

                      query_id               class_category     rr      time  
8212                mech_label           instances_100_1000  1.000  1.003011  
8213                 mech_desc           instances_100_1000  1.000  1.091792  
8214           mech_label_desc           instances_100_1000  1.000  1.197138  
8215   mech_label_user_query_0           instances_100_1000  1.000  1.035266  
8216   mech_label_user_query_1           instances_100_1000  1.000  1.058481  
...                        ...                          ...    ...       ...  
15075  mech_label_user_query_1  instances_100000_1000000000  1.000  1.166308  
15076  mech_label_user_query_2  instances_100000_1000000000  1.000  1.448362  
15077  mech_label_user_query_3  instances_100000_1000000000  0.200  1.218570  
15078  mech_label_user_query_4  instances_100000_1000000000  0.125  1.189481  
15079  mech_label_user_query_5  instances_100000_1000000000  0.500  1.167621  

[4180 rows x 7 columns]

In [14]:
print(df_only_mech_queries[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())


                                                                                       method_id        rr      time
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.872445  1.343595
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.871401  1.268795
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.871336  1.359918
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.870785  1.367068
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.741906  1.405028


### Only labels

In [15]:
df_only_mech_queries_labels_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_label"])]
df_only_mech_queries_labels_only

method_id  user_id  class_id  \
8212         single_elastic_bm25_boost_0.5_cross_encoder        0  20825628   
8221         single_elastic_bm25_boost_0.5_cross_encoder        0   2066754   
8230         single_elastic_bm25_boost_0.5_cross_encoder        0     34038   
8239         single_elastic_bm25_boost_0.5_cross_encoder        0   7725634   
8248         single_elastic_bm25_boost_0.5_cross_encoder        0    602884   
...                                                  ...      ...       ...   
14987  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0   8016240   
14996  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     23442   
15005  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     30022   
15014  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0    130019   
15023  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0  28942331   

         query_id               class_category   rr      time  
8212   mech_label           instances_100_1000  1.0  1.003011  
8221   mech_label         instances_1000_10000  1.0  0.977809  
8230   mech_label       instances_10000_100000  1.0  0.842765  
8239   mech_label  instances_100000_1000000000  1.0  1.108354  
8248   mech_label        children_10000_100000  1.0  0.947576  
...           ...                          ...  ...       ...  
14987  mech_label           instances_100_1000  0.5  1.655483  
14996  mech_label  instances_100000_1000000000  1.0  1.074946  
15005  mech_label         instances_1000_10000  0.5  0.838135  
15014  mech_label       instances_10000_100000  1.0  1.167742  
15023  mech_label           instances_100_1000  1.0  1.265985  

[280 rows x 7 columns]

In [16]:
print(df_only_mech_queries_labels_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                                       method_id        rr      time
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.916802  1.128742
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.909858  1.133374
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.909510  1.082071
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.909510  1.163659
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.909425  1.154104


### Only descriptions

In [17]:
df_only_mech_queries_desc_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_desc"])]
df_only_mech_queries_desc_only

method_id  user_id  class_id  \
8213         single_elastic_bm25_boost_0.5_cross_encoder        0  20825628   
8222         single_elastic_bm25_boost_0.5_cross_encoder        0   2066754   
8231         single_elastic_bm25_boost_0.5_cross_encoder        0     34038   
8240         single_elastic_bm25_boost_0.5_cross_encoder        0   7725634   
8249         single_elastic_bm25_boost_0.5_cross_encoder        0    602884   
...                                                  ...      ...       ...   
14988  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0   8016240   
14997  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     23442   
15006  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     30022   
15015  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0    130019   
15024  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0  28942331   

        query_id               class_category   rr      time  
8213   mech_desc           instances_100_1000  1.0  1.091792  
8222   mech_desc         instances_1000_10000  1.0  1.370641  
8231   mech_desc       instances_10000_100000  0.0  1.574008  
8240   mech_desc  instances_100000_1000000000  1.0  1.234325  
8249   mech_desc        children_10000_100000  1.0  3.174682  
...          ...                          ...  ...       ...  
14988  mech_desc           instances_100_1000  1.0  1.095909  
14997  mech_desc  instances_100000_1000000000  1.0  1.484046  
15006  mech_desc         instances_1000_10000  1.0  1.205881  
15015  mech_desc       instances_10000_100000  1.0  1.251892  
15024  mech_desc           instances_100_1000  0.5  1.298140  

[270 rows x 7 columns]

In [18]:
print(df_only_mech_queries_desc_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                                       method_id        rr      time
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.925926  1.395361
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.925926  1.302648
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.925926  1.399885
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.925926  1.412063
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.472222  1.569442


### Only labels with descriptions

In [19]:
df_only_mech_queries_label_desc_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_label_desc"])]
df_only_mech_queries_label_desc_only

method_id  user_id  class_id  \
8214         single_elastic_bm25_boost_0.5_cross_encoder        0  20825628   
8223         single_elastic_bm25_boost_0.5_cross_encoder        0   2066754   
8232         single_elastic_bm25_boost_0.5_cross_encoder        0     34038   
8241         single_elastic_bm25_boost_0.5_cross_encoder        0   7725634   
8250         single_elastic_bm25_boost_0.5_cross_encoder        0    602884   
...                                                  ...      ...       ...   
14989  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0   8016240   
14998  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     23442   
15007  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     30022   
15016  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0    130019   
15025  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0  28942331   

              query_id               class_category   rr      time  
8214   mech_label_desc           instances_100_1000  1.0  1.197138  
8223   mech_label_desc         instances_1000_10000  1.0  1.092034  
8232   mech_label_desc       instances_10000_100000  1.0  1.591044  
8241   mech_label_desc  instances_100000_1000000000  1.0  1.295824  
8250   mech_label_desc        children_10000_100000  1.0  1.193397  
...                ...                          ...  ...       ...  
14989  mech_label_desc           instances_100_1000  1.0  1.150884  
14998  mech_label_desc  instances_100000_1000000000  1.0  1.250597  
15007  mech_label_desc         instances_1000_10000  1.0  1.205848  
15016  mech_label_desc       instances_10000_100000  1.0  1.268469  
15025  mech_label_desc           instances_100_1000  1.0  1.411801  

[270 rows x 7 columns]

In [20]:
print(df_only_mech_queries_label_desc_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                                       method_id        rr      time
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  1.000000  1.413391
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  1.000000  1.299404
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  1.000000  1.421032
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  1.000000  1.417457
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.888889  1.553862


### Only labels with user queries

In [21]:
df_only_mech_queries_label_user_only = df_only_mech_queries[~df_only_mech_queries["query_id"].isin(["mech_desc", "mech_label", "mech_label_desc"])]
df_only_mech_queries_label_user_only

method_id  user_id  class_id  \
8215         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
8216         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
8217         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
8218         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
8219         single_elastic_bm25_boost_0.5_cross_encoder      101  20825628   
...                                                  ...      ...       ...   
15075  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15076  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15077  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15078  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
15079  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   

                      query_id               class_category     rr      time  
8215   mech_label_user_query_0           instances_100_1000  1.000  1.035266  
8216   mech_label_user_query_1           instances_100_1000  1.000  1.058481  
8217   mech_label_user_query_2           instances_100_1000  0.500  1.240887  
8218   mech_label_user_query_3           instances_100_1000  1.000  1.248775  
8219   mech_label_user_query_4           instances_100_1000  1.000  1.094274  
...                        ...                          ...    ...       ...  
15075  mech_label_user_query_1  instances_100000_1000000000  1.000  1.166308  
15076  mech_label_user_query_2  instances_100000_1000000000  1.000  1.448362  
15077  mech_label_user_query_3  instances_100000_1000000000  0.200  1.218570  
15078  mech_label_user_query_4  instances_100000_1000000000  0.125  1.189481  
15079  mech_label_user_query_5  instances_100000_1000000000  0.500  1.167621  

[3360 rows x 7 columns]

In [22]:
print(df_only_mech_queries_label_user_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                                       method_id        rr      time
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.854780  1.351345
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.853510  1.279175
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.853436  1.368947
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.852744  1.376354
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.737192  1.402880


### Overall per category

In [23]:
df_only_mech_queries_per_category_thin = df_only_mech_queries[["method_id", "class_category", "rr", "time"]]
df_only_mech_queries_per_category_thin

method_id  \
8212         single_elastic_bm25_boost_0.5_cross_encoder   
8213         single_elastic_bm25_boost_0.5_cross_encoder   
8214         single_elastic_bm25_boost_0.5_cross_encoder   
8215         single_elastic_bm25_boost_0.5_cross_encoder   
8216         single_elastic_bm25_boost_0.5_cross_encoder   
...                                                  ...   
15075  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15076  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15077  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15078  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
15079  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   

                    class_category     rr      time  
8212            instances_100_1000  1.000  1.003011  
8213            instances_100_1000  1.000  1.091792  
8214            instances_100_1000  1.000  1.197138  
8215            instances_100_1000  1.000  1.035266  
8216            instances_100_1000  1.000  1.058481  
...                            ...    ...       ...  
15075  instances_100000_1000000000  1.000  1.166308  
15076  instances_100000_1000000000  1.000  1.448362  
15077  instances_100000_1000000000  0.200  1.218570  
15078  instances_100000_1000000000  0.125  1.189481  
15079  instances_100000_1000000000  0.500  1.167621  

[4180 rows x 4 columns]

In [24]:
df_only_mech_queries_overall_per_category = df_only_mech_queries_per_category_thin.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category", "rr"], ascending=False)
print(df_only_mech_queries_overall_per_category.to_string())

                                                                                        method_id               class_category        rr      time
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder           instances_100_1000  0.850520  1.334776
15                                                    single_qdrant_dense_boost_0.5_cross_encoder           instances_100_1000  0.849517  1.301341
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder           instances_100_1000  0.847943  1.329151
23                                                   single_qdrant_sparse_boost_0.5_cross_encoder           instances_100_1000  0.842189  1.248103
7                                                     single_elastic_bm25_boost_0.5_cross_encoder           instances_100_1000  0.776574  1.399823
14                                                    single_qdrant_dense_boost_0.5_cross_encoder         instances_10

# Overall - only user queries

In [25]:
df_only_user_queries = df[df["query_id"].str.startswith("user")]
df_only_user_queries

method_id  user_id  class_id  \
7540         single_elastic_bm25_boost_0.5_cross_encoder        1  20825628   
7541         single_elastic_bm25_boost_0.5_cross_encoder        1  20825628   
7542         single_elastic_bm25_boost_0.5_cross_encoder        1  20825628   
7543         single_elastic_bm25_boost_0.5_cross_encoder        1  20825628   
7544         single_elastic_bm25_boost_0.5_cross_encoder        1  20825628   
...                                                  ...      ...       ...   
14239  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
14240  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
14241  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
14242  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
14243  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   

           query_id               class_category        rr      time  
7540   user_query_0           instances_100_1000  0.000000  1.079723  
7541   user_query_1           instances_100_1000  0.000000  1.087204  
7542   user_query_2           instances_100_1000  0.000000  1.295412  
7543   user_query_3           instances_100_1000  0.000000  1.191521  
7544   user_query_4           instances_100_1000  0.000000  0.971411  
...             ...                          ...       ...       ...  
14239  user_query_1  instances_100000_1000000000  0.500000  2.313716  
14240  user_query_2  instances_100000_1000000000  0.500000  1.432357  
14241  user_query_3  instances_100000_1000000000  0.000000  1.144615  
14242  user_query_4  instances_100000_1000000000  0.000000  1.349599  
14243  user_query_5  instances_100000_1000000000  0.090909  1.130753  

[3360 rows x 7 columns]

In [26]:
print(df_only_user_queries[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                                       method_id        rr      time
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.309535  1.305971
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.307532  1.310053
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.307372  1.311503
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.293274  1.233126
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.168382  1.289671


### Overall per category

In [27]:
df_only_user_queries_per_category_thin = df_only_user_queries[["method_id", "class_category", "rr", "time"]]
df_only_user_queries_per_category_thin

method_id  \
7540         single_elastic_bm25_boost_0.5_cross_encoder   
7541         single_elastic_bm25_boost_0.5_cross_encoder   
7542         single_elastic_bm25_boost_0.5_cross_encoder   
7543         single_elastic_bm25_boost_0.5_cross_encoder   
7544         single_elastic_bm25_boost_0.5_cross_encoder   
...                                                  ...   
14239  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
14240  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
14241  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
14242  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
14243  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   

                    class_category        rr      time  
7540            instances_100_1000  0.000000  1.079723  
7541            instances_100_1000  0.000000  1.087204  
7542            instances_100_1000  0.000000  1.295412  
7543            instances_100_1000  0.000000  1.191521  
7544            instances_100_1000  0.000000  0.971411  
...                            ...       ...       ...  
14239  instances_100000_1000000000  0.500000  2.313716  
14240  instances_100000_1000000000  0.500000  1.432357  
14241  instances_100000_1000000000  0.000000  1.144615  
14242  instances_100000_1000000000  0.000000  1.349599  
14243  instances_100000_1000000000  0.090909  1.130753  

[3360 rows x 4 columns]

In [28]:
df_only_user_queries_overall_per_category = df_only_user_queries_per_category_thin.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category", "rr"], ascending=False)
print(df_only_user_queries_overall_per_category.to_string())

                                                                                        method_id               class_category        rr      time
23                                                   single_qdrant_sparse_boost_0.5_cross_encoder           instances_100_1000  0.309279  1.211957
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder           instances_100_1000  0.305733  1.279618
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder           instances_100_1000  0.301768  1.291400
15                                                    single_qdrant_dense_boost_0.5_cross_encoder           instances_100_1000  0.301284  1.306490
7                                                     single_elastic_bm25_boost_0.5_cross_encoder           instances_100_1000  0.169337  1.266422
14                                                    single_qdrant_dense_boost_0.5_cross_encoder         instances_10

### Choosing the best match per class of a user

- The idea is that if user writes the 6 queries, and at least one query find the result, the user can use it to model the data structure.
- We want to choose the best query for each of the user assigned classes.
- We can do it by grouping by: method, user, and class.
  - Then we select a maximum for the groups from reciprocal rank.
  - Then the results will be used to again group by a method. 

In [29]:
df_only_user_queries_max_per_user_class = df_only_user_queries[["method_id", "user_id", "class_id", "rr"]].groupby(["method_id", "user_id", "class_id"], as_index=False).max()
df_only_user_queries_max_per_user_class

method_id  user_id  class_id  \
0          single_elastic_bm25_boost_0.5_cross_encoder        1     34038   
1          single_elastic_bm25_boost_0.5_cross_encoder        1    602884   
2          single_elastic_bm25_boost_0.5_cross_encoder        1   2066754   
3          single_elastic_bm25_boost_0.5_cross_encoder        1   4897819   
4          single_elastic_bm25_boost_0.5_cross_encoder        1   7725634   
..                                                 ...      ...       ...   
555  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14   8016240   
556  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  11483816   
557  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  18593264   
558  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  28942331   
559  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  51591359   

           rr  
0    0.000000  
1    0.000000  
2    0.000000  
3    0.000000  
4    0.000000  
..        ...  
555  0.333333  
556  1.000000  
557  1.000000  
558  0.200000  
559  1.000000  

[560 rows x 4 columns]

### Average rr with only best ranks per classes of users


In [30]:
print(df_only_user_queries_max_per_user_class[["method_id", "rr"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                                       method_id        rr
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  0.695311
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  0.689860
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  0.684880
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  0.680145
0                                                    single_elastic_bm25_boost_0.5_cross_encoder  0.387202


### Number of classes found per method with only best ranks

- In total we have 56 classes and each class is given to two users.
- Each method have a change of scoring 112 found classes.

In [31]:
df_only_user_queries_max_per_user_class_thin = df_only_user_queries_max_per_user_class[["method_id", "rr"]].copy()
# Round up rr meaning the method found the class.
df_only_user_queries_max_per_user_class_thin["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
# Group the results by method -> for a method, each class is presented twice (since two users provided the queries).
print(df_only_user_queries_max_per_user_class_thin.groupby(["method_id"], as_index=False).sum().sort_values(by="rr", ascending=False).to_string())


                                                                                       method_id   rr
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder  105
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder  105
2                                                   single_qdrant_sparse_boost_0.5_cross_encoder  103
1                                                    single_qdrant_dense_boost_0.5_cross_encoder  102
0                                                    single_elastic_bm25_boost_0.5_cross_encoder   50


### For each method and a class we get the number of successful search queries

- There is 12 tries for each method, since we got 2 users and 6 queries per a user.


In [32]:
df1_method_class_successes_thin = df_only_user_queries[["method_id", "class_id", "rr"]]
df1_method_class_successes_thin = df1_method_class_successes_thin.copy()
df1_method_class_successes_thin["rr"] = df1_method_class_successes_thin["rr"].apply(lambda x: math.ceil(x))
print(df1_method_class_successes_thin.groupby(["class_id", "method_id"], as_index=False).sum().to_string())

      class_id                                                                                      method_id  rr
0          532                                                    single_elastic_bm25_boost_0.5_cross_encoder   0
1          532                                                    single_qdrant_dense_boost_0.5_cross_encoder   7
2          532                                                   single_qdrant_sparse_boost_0.5_cross_encoder   7
3          532  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_boost_0.5_cross_encoder   7
4          532                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_boost_0.5_cross_encoder   7
5        11303                                                    single_elastic_bm25_boost_0.5_cross_encoder   2
6        11303                                                    single_qdrant_dense_boost_0.5_cross_encoder   9
7        11303                                                   single_qdrant_sparse_bo

- If we sum this per a class
- We have 66 methods (3 base, 27 tuple, 36 triple) * 3 (for each boost weight), the class appears twice and 6 queries per class.

In [33]:
print(df1_method_class_successes_thin[["class_id", "rr"]].groupby(["class_id"], as_index=False).sum().sort_values(by="rr", ascending=False).to_string())

     class_id  rr
36   11483816  55
29    3007842  53
46   23002039  53
48   29654788  50
12     130019  50
38   13442814  49
16     294414  49
39   15052846  48
25    1478437  46
28    2819161  45
22    1348305  45
24    1371150  44
49   30325611  44
51   51591359  42
42   19389637  41
1       11303  41
41   18593264  39
20    1134686  38
45   21191270  37
19    1076052  37
23    1364302  37
27    2319498  36
50   35922124  36
4       30022  36
3       24856  35
47   28942331  34
6       34763  34
55  124030970  33
33    5783996  32
17     310899  32
8       44539  32
9       47521  32
37   12890393  31
52   63125083  30
15     179049  30
10      79007  29
0         532  28
30    3231690  28
43   20755546  28
31    3621673  27
53  123523193  27
5       34038  27
18     602884  26
2       23442  24
26    2066754  20
14     134556  20
7       43229  19
44   20825628  18
21    1195098  15
34    7725634  15
40   18557321  14
11     124714  12
35    8016240  11
13     133036  10
32    4897

### Classes the best methods did not find from no rerank results

- triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1
- tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6

In [34]:
# Triple fusion triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1

df1_triple_best = df_only_user_queries_max_per_user_class[df_only_user_queries_max_per_user_class["method_id"].str.contains("triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1")][["method_id", "class_id", "rr"]].copy()
df1_triple_best["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
df1_triple_best[df1_triple_best["rr"] == 0]

method_id   class_id  rr
339  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    4897819   0
340  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    7725634   0
351  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...  123691918   0
355  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    4897819   0
365  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   20755546   0
367  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...  123691918   0
405  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    1195098   0

- The class it did not find at all:
    - 4897819 - role
    - 123691918 - tool

In [35]:
# Tuple fusion tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6

df1_triple_best = df_only_user_queries_max_per_user_class[df_only_user_queries_max_per_user_class["method_id"].str.contains("tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6")][["method_id", "class_id", "rr"]].copy()
df1_triple_best["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
df1_triple_best[df1_triple_best["rr"] == 0]

method_id   class_id  rr
451  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    4897819   0
452  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    7725634   0
463  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...  123691918   0
467  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    4897819   0
477  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...   20755546   0
479  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...  123691918   0
517  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    1195098   0

- The classes it did not find at all
    - 4897819 - role
    - 123691918 - tool

Interestingly, they are the same as before. Looking into the queries, we tried them manually, and it did not find them.
Thinking, maybe it was since the things were too abstract, the users have a big problem with generating the queries to.

# The least found classes

In [36]:
df[["class_id", "class_category", "rr"]].groupby(["class_id", "class_category"], as_index=False).mean().sort_values(by="rr", ascending=False)

class_id               class_category        rr
12     130019       instances_10000_100000  0.882434
39   15052846           instances_100_1000  0.871869
24    1371150           instances_100_1000  0.800029
51   51591359       instances_10000_100000  0.786420
25    1478437         instances_1000_10000  0.782669
28    2819161           instances_100_1000  0.778768
45   21191270  instances_100000_1000000000  0.764887
48   29654788  instances_100000_1000000000  0.762469
49   30325611            ancestors_150_180  0.749630
43   20755546     ancestors_180_1000000000  0.740741
36   11483816         instances_1000_10000  0.734654
19    1076052         instances_1000_10000  0.732540
27    2319498       instances_10000_100000  0.718815
50   35922124           instances_100_1000  0.710907
22    1348305       instances_10000_100000  0.706772
29    3007842         instances_1000_10000  0.705455
46   23002039         instances_1000_10000  0.702941
38   13442814  instances_100000_1000000000  0.697910
16     294414        children_10000_100000  0.693746
41   18593264  instances_100000_1000000000  0.693531
20    1134686       instances_10000_100000  0.691685
17     310899         instances_1000_10000  0.687695
55  124030970     ancestors_180_1000000000  0.686067
23    1364302           instances_100_1000  0.682403
3       24856         instances_1000_10000  0.680612
52   63125083           instances_100_1000  0.678054
42   19389637  instances_100000_1000000000  0.664501
5       34038       instances_10000_100000  0.636476
4       30022         instances_1000_10000  0.636430
1       11303         instances_1000_10000  0.620947
26    2066754         instances_1000_10000  0.610018
47   28942331           instances_100_1000  0.594092
6       34763       instances_10000_100000  0.589174
15     179049       instances_10000_100000  0.588519
34    7725634  instances_100000_1000000000  0.566572
44   20825628           instances_100_1000  0.556311
9       47521  instances_100000_1000000000  0.541587
0         532  instances_100000_1000000000  0.535036
30    3231690       instances_10000_100000  0.531779
40   18557321            ancestors_150_180  0.526861
10      79007  instances_100000_1000000000  0.506625
2       23442  instances_100000_1000000000  0.487239
31    3621673         instances_1000_10000  0.480937
7       43229  instances_100000_1000000000  0.473947
33    5783996       instances_10000_100000  0.473432
8       44539         instances_1000_10000  0.456049
37   12890393           instances_100_1000  0.450460
18     602884        children_10000_100000  0.418190
13     133036           instances_100_1000  0.350394
21    1195098       instances_10000_100000  0.348865
35    8016240           instances_100_1000  0.344756
11     124714       instances_10000_100000  0.268880
54  123691918   children_100000_1000000000  0.264274
14     134556  instances_100000_1000000000  0.221153
53  123523193           instances_100_1000  0.217509
32    4897819   children_100000_1000000000  0.082906